kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java

```diff
@@ -0,0 +1,285 @@
+        String resourceVersion = metadata.getResourceVersion();
+        if (!Utils.StringUtils.isNullOrEmpty(resourceVersion)) {
+          return resourceVersion;
+        }
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Returns the kind of the entity
+   */
+  public static String getKind(HasMetadata entity) {
+    if (entity != null) {
+      // TODO use reflection to find the kind?
+      if (entity instanceof KubernetesList) {
+        return "List";
+      } else {
+        return entity.getClass().getSimpleName();
+      }
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Returns Qualified name for the specified Kubernetes Resource
+   *
+   * @param entity Kubernetes resource
+   * @return
+   */
+  public static String getQualifiedName(HasMetadata entity) {
+    if (entity != null) {
+      return "" + getNamespace(entity) + "/" + getName(entity);
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Returns name of the resource from it's Metadata
+   *
+   * @param entity Kubernetes resource
+   * @return
+   */
+  public static String getName(HasMetadata entity) {
+    if (entity != null) {
+      return getName(entity.getMetadata());
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Returns true if this entity has a valid non blank resourceVersion in its metadata
+   */
+  public static boolean hasResourceVersion(HasMetadata entity) {
+    return getResourceVersion(entity) != null;
+  }
+
+  /**
+   * Returns name of the resource from it's Metadata
+   *
+   * @param entity MetaData of kubernetes resource
+   * @return
+   */
+  public static String getName(ObjectMeta entity) {
+    if (entity != null) {
+      return Utils.StringUtils.coalesce(entity.getName(),
+        getAdditionalPropertyText(entity.getAdditionalProperties(), "id"),
+        entity.getUid());
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Used to get additional properties from Object's metadata
+   *
+   * @param additionalProperties
+   * @param name
+   * @return
+   */
+  protected static String getAdditionalPropertyText(Map<String, Object> additionalProperties, String name) {
+    if (additionalProperties != null) {
+      Object value = additionalProperties.get(name);
+      if (value != null) {
+        return value.toString();
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Null safe set operation for setting Kubernetes resource's name
+   *
+   * @param entity Kubernetes Resource
+   * @param name name to be set
+   */
+  public static void setName(HasMetadata entity, String name) {
+    getOrCreateMetadata(entity).setName(name);
+  }
+
+  /**
+   * Null safe set operation of setting Kubernetes resource's name/namespace
+   *
+   * @param entity Kubernetes
+   * @param namespace
+   * @param name
+   */
+  public static void setName(HasMetadata entity, String namespace, String name) {
+    ObjectMeta metadata = getOrCreateMetadata(entity);
+    metadata.setNamespace(namespace);
```

Comment:
I'd like to avoid bringing in utils that mutate resources. IMHO resources should be immutable and all manipulations should happen via builders. It may be too late for that now, as it's too disruptive, but at least we should avoid doing that internally.

So either have setXXX use builders and return a new object or remove them all together.